src/transformers/models/jetmoe/modeling_jetmoe.py

```diff
@@ -0,0 +1,1731 @@
+
+    def forward(self, inputs, expert_size):
+        """
+        Forward pass of the JetMoeParallelExperts module.
+
+        Args:
+            inputs (Tensor):
+                Input tensor.
+            expert_size:
+                Expert size information.
+
+        Returns:
+            Tensor: Output tensor.
+        """
+        input_list = inputs.split(expert_size, dim=0)
+        output_list = []
+        for i in range(self.num_experts):
+            output_list.append(F.linear(input_list[i], self.weight[i]))
+        results = torch.cat(output_list, dim=0)
+        return results
+
+
+class JetMoeTopKGating(nn.Module):
+    def __init__(self, input_size: int, num_experts: int, top_k: int):
+        """
+        Initialize the top-k gating mechanism.
+
+        Args:
+            input_size (`int`):
+                Size of the input.
+            num_experts (`int`):
+                Number of experts.
+            top_k (`int`):
+                Number of top experts to select.
+        """
+        super().__init__()
+
+        self.num_experts = num_experts
+        self.input_size = input_size
+        self.top_k = top_k
+
+        self.layer = nn.Linear(input_size, num_experts, bias=False)
+
+    def extra_repr(self):
+        """
+        Return extra representation string for the module.
+        """
+        return "k={}, num_experts={}".format(self.top_k, self.num_experts)
+
+    def compute_topo(self, top_k_indices, top_k_gates):
+        zeros = torch.zeros(
+            [top_k_gates.size(0), self.num_experts], dtype=top_k_gates.dtype, device=top_k_gates.device
+        )
+        gates = zeros.scatter(1, top_k_indices, 1)
+        expert_size = gates.long().sum(0)
+        top_k_gates = top_k_gates.flatten()
+        top_k_experts = top_k_indices.flatten()
+        _, index_sorted_experts = top_k_experts.sort(0)
+        batch_index = index_sorted_experts.div(self.top_k, rounding_mode="trunc")
+        batch_gates = top_k_gates[index_sorted_experts]
+        expert_size = expert_size.tolist()
+
+        return index_sorted_experts, batch_index, batch_gates, expert_size
+
+    def forward(self, x, return_topo=False):
+        """
+        Compute the top-k gating for the input.
+
+        See paper: https://arxiv.org/abs/1701.06538.
+
+        Args:
+            x (torch.Tensor):
+                Input tensor with shape [batch_size, input_size].
+
+        Returns:
+            torch.Tensor:
+                Top-k indices.
+            torch.Tensor:
+                Top-k gating values.
+            torch.Tensor:
+                router layer logits.
+        """
+
+        logits = self.layer(x).float()
+        top_k_logits, top_k_indices = logits.topk(self.top_k, dim=1)
+        top_k_gates = torch.softmax(top_k_logits, dim=1).type_as(x)
+
+        if return_topo:
+            return self.compute_topo(top_k_indices, top_k_gates) + (logits,)
+        else:
+            return top_k_indices, top_k_gates, logits
```

Comment:
```suggestion
        outputs = self.compute_topo(top_k_indices, top_k_gates) + (logits,)
        return outputs
```
See my comments above